staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,328 @@
+	wcib := &watchCacheIntervalBuffer{}
+	wcib.buffer = make([]*watchCacheEvent, len(events))
+	copy(wcib.buffer, events)
+	wcib.endIndex = len(events)
+	return wcib
+}
+
+func TestIntervalBufferIsFull(t *testing.T) {
+	cases := []struct {
+		endIndex int
+		expected bool
+	}{
+		{bufferSize - 1, false},
+		{bufferSize, true},
+		{bufferSize + 1, true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{endIndex: c.endIndex}
+		actual := wcib.isFull()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferIsEmpty(t *testing.T) {
+	cases := []struct {
+		startIndex int
+		endIndex   int
+		expected   bool
+	}{
+		{0, 10, false},
+		{50, 50, true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{
+			startIndex: c.startIndex,
+			endIndex:   c.endIndex,
+		}
+		actual := wcib.isEmpty()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferNext(t *testing.T) {
+	events := []*watchCacheEvent{
+		{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+		{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+		{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+	}
+	wcib := bufferFromEvents(events)
+
+	for _, e := range events {
+		event, ok := wcib.next()
+		if !ok {
+			t.Error("unexpected bool value indicating empty buffer")
+		}
+		if !reflect.DeepEqual(event, e) {
+			t.Errorf("unexpected event, expected: %v, got %v", *event, *e)
+		}
+	}
+
+	event, ok := wcib.next()
+	if ok {
+		t.Error("unexpected bool value indicating buffer is not empty")
+	}
+	if event != nil {
+		t.Errorf("unexpected event received, expected: nil, got %v", *event)
+	}
+}
+
+func TestFillBuffer(t *testing.T) {
+	events := []*watchCacheEvent{
+		{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+		{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+		{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+	}
+	wci := intervalFromEvents(events)
+	wci.fillBuffer()
+	for i := 0; i < len(events); i++ {
+		if !reflect.DeepEqual(events[i], wci.buffer.buffer[i]) {
+			t.Errorf("expected: %v, got: %v", *events[i], *wci.buffer.buffer[i])
+		}
+	}
+}
+
+func TestCacheIntervalFromWatchCache(t *testing.T) {
+	// 10 is arbitrary, the point is to have
+	// capacity more than the size of the cache
+	// interval's buffer.
+	const capacity = bufferSize + 10
+	wc := newTestWatchCache(capacity, &cache.Indexers{})
+	for i := 0; i < capacity; i++ {
+		wc.Add(makeTestPod(
+			fmt.Sprintf("pod%d", i+1),
+			uint64(i+1),
+		))
+	}
+
+	indexerFunc := func(i int) *watchCacheEvent {
+		return wc.cache[i%wc.capacity]
+	}
+	// 5 is again arbitrary, point is to have endIndex
+	// cross the cache interval's buffer size but not
+	// span the entire watch cache.
+	wci := newCacheInterval(
+		0,
+		bufferSize+5,
+		indexerFunc,
+		wc.isIndexValid,
+		&wc.RWMutex,
+	)
+
+	event, ok, err := wci.Next()
```

Comment:
Let's do this differently:
- let's keep the expectedEvents slice as part of testcase

And here just do:
```
receives := 0
for {
  isEmpty := wci.buffer.isEmpty()
  if isEmpty != (received%bufferSize ==0) {
    t.Errrof("...)
  }

  event, ok, err := wci.Next()
  if err != nil {
    ...
  }
  if !ok {
    if receive<len(expected) {
     t.Errrof(...)
    }
    break
  }
  if !reflect.DeepEqual(event, expected[received]) {
    t.Errorf(...)
  }
  received++
}
```

This code will be much simpler